sdk/src/main/java/io/opentelemetry/sdk/metrics/BaseInstrument.java

```diff
@@ -0,0 +1,161 @@
+import java.util.Map;
+
+public abstract class BaseInstrument<B> implements Instrument<B> {
+
+  private final Map<LabelSet, B> boundCounters = new HashMap<>();
+
+  private final String name;
+  private final String description;
+  private final Map<String, String> constantLabels;
+  private final List<String> labelKeys;
+
+  protected BaseInstrument(
+      String name, String description, Map<String, String> constantLabels, List<String> labelKeys) {
+    this.name = name;
+    this.description = description;
+    this.constantLabels = constantLabels;
+    this.labelKeys = labelKeys;
+  }
+
+  @Override
+  public B bind(LabelSet labelSet) {
+    return getOrCreate(labelSet);
+  }
+
+  private B getOrCreate(LabelSet labelSet) {
+    B boundLongCounter = boundCounters.get(labelSet);
+    if (boundLongCounter == null) {
+      boundLongCounter = create(labelSet);
+      boundCounters.put(labelSet, boundLongCounter);
+    }
+    return boundLongCounter;
+  }
+
+  protected abstract B create(LabelSet labelSet);
+
+  @SuppressWarnings("rawtypes")
+  @Override
+  public void unbind(B boundInstrument) {
+    boundCounters.remove(((BaseBoundInstrument) boundInstrument).labels);
+  }
+
+  @Override
+  public String toString() {
+    return "BaseInstrument{"
+        + "boundCounters="
+        + boundCounters
+        + ", name='"
+        + name
+        + '\''
+        + ", description='"
+        + description
+        + '\''
+        + ", constantLabels="
+        + constantLabels
+        + ", labelKeys="
+        + labelKeys
+        + '}';
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (!(o instanceof BaseInstrument)) {
+      return false;
+    }
+
+    BaseInstrument<?> that = (BaseInstrument<?>) o;
+
+    if (boundCounters != null
+        ? !boundCounters.equals(that.boundCounters)
+        : that.boundCounters != null) {
+      return false;
+    }
+    if (name != null ? !name.equals(that.name) : that.name != null) {
+      return false;
+    }
+    if (description != null ? !description.equals(that.description) : that.description != null) {
+      return false;
+    }
+    if (constantLabels != null
+        ? !constantLabels.equals(that.constantLabels)
+        : that.constantLabels != null) {
+      return false;
+    }
+    return labelKeys != null ? labelKeys.equals(that.labelKeys) : that.labelKeys == null;
+  }
+
+  @Override
+  public int hashCode() {
+    int result = boundCounters != null ? boundCounters.hashCode() : 0;
+    result = 31 * result + (name != null ? name.hashCode() : 0);
+    result = 31 * result + (description != null ? description.hashCode() : 0);
+    result = 31 * result + (constantLabels != null ? constantLabels.hashCode() : 0);
+    result = 31 * result + (labelKeys != null ? labelKeys.hashCode() : 0);
+    return result;
+  }
+
+  static class BaseBoundInstrument<I extends BaseInstrument<?>> {
+
+    private final I instrument;
```

Comment:
Not sure you need the instrument here, you will attach the aggregations that belong to the instrument directly. So I would suggest to just remove it for the moment.

Reply:
Hmm. How would you tie the bound instrument together with the instrument that it is binding, then?

Reply:
By sharing a deeper state (Aggregation) that belongs to the instrument, and not the instrument itself.

Reply:
that makes sense. I gutted this and replaced with todos to wire up to an aggregator/accumulator